reactor-netty-core/src/main/java/reactor/netty/logging/ReactorNettyLoggingHandler.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2011-Present VMware, Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.netty.logging;
+
+import java.nio.charset.Charset;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufHolder;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.logging.LogLevel;
+import io.netty.handler.logging.LoggingHandler;
+
+/**
+ * Extends {@link LoggingHandler} and logs all events in a
+ * textual representation so it's human readable and less verbose.
+ *
+ * Hint: Logger escapes newlines as "\n" to reduce output.
+ */
+public class ReactorNettyLoggingHandler extends LoggingHandler {
+
+	private final Charset charset;
+
+	/**
+	 * Creates a new instance with the specified logger name and charset.
+	 *
+	 * @param name    the name of the class to use for the logger
+	 * @param level   the log level
+	 * @param charset the charset used to decode the ByteBuf
+	 */
+	public ReactorNettyLoggingHandler(final String name, final LogLevel level, final Charset charset) {
+		super(name, level);
+		this.charset = charset;
+	}
+
+	@Override
+	protected String format(ChannelHandlerContext ctx, String eventName, Object arg) {
+		if (arg instanceof ByteBuf) {
+			return formatByteBuf(ctx, eventName, (ByteBuf) arg);
+		}
+		else if (arg instanceof ByteBufHolder) {
+			return formatByteBuf(ctx, eventName, ((ByteBufHolder) arg).content());
+		}
+		else {
+			return super.format(ctx, eventName, arg);
+		}
+	}
+
+	/* Format of message is:
+	 * CHANNEL EVENT_NAME: MESSAGE_LENGTH_IN_BYTESB CONTENT_OF_BYTE_BUF
+	 *
+	 * == Example ==
+	 * Channel: [id: 0x329c6ffd, L:/127.0.0.1:52640 - R:/127.0.0.1:8080]
+	 * EVENT_NAME: WRITE
+	 * MESSAGE_LENGTH_IN_BYTES: 4
+	 * CONTENT_OF_BYTE_BUF: echo
+	 *
+	 * result: [id: 0x329c6ffd, L:/127.0.0.1:52640 - R:/127.0.0.1:8080] WRITE: 4B echo
+	 */
+	private String formatByteBuf(ChannelHandlerContext ctx, String eventName, ByteBuf msg) {
+		final String chStr = ctx.channel().toString();
+		final String message = msg.toString(charset).replace(System.lineSeparator(), "\\n");
```

Comment:
I think this line needs a bit more attention. 

The code is currently using `msg.toString(charset)` to get the String representation of the ByteBuf object. I found [this](https://github.com/netty/netty/blob/00afb19d7a37de21b35ce4f6cb3fa7f74809f2ab/buffer/src/main/java/io/netty/buffer/ByteBuf.java#L237) in the ByteBuf documentation and I'm not sure what they mean with 

> is not a conversion method.

Does this mean I need to write a separate method that converts the ByteBuf to a String?

Back to the line separator topic. I think adding `.replace("\n", "\\n").replace("\r", "\\r")` should do the job but I'm not sure if this is the best solution to solve it.

Reply:
> The code is currently using `msg.toString(charset)` to get the String representation of the ByteBuf object. I found [this](https://github.com/netty/netty/blob/00afb19d7a37de21b35ce4f6cb3fa7f74809f2ab/buffer/src/main/java/io/netty/buffer/ByteBuf.java#L237) in the ByteBuf documentation and I'm not sure what they mean with
> 
> > is not a conversion method.
> 
> Does this mean I need to write a separate method that converts the ByteBuf to a String?

Interesting
Even `io.netty.buffer.ByteBuf#readCharSequence` calls `toString` the difference is that `readCharSequence` changes the read index and `toString` does not.
Both methods end up calling `io.netty.buffer.ByteBufUtil#decodeString`

Reply:
> Back to the line separator topic. I think adding `.replace("\n", "\\n").replace("\r", "\\r")` should do the job but I'm not sure if this is the best solution to solve it.

I'm also wondering... With the current implementation we will have in the log this

```
09:36:59.056 [reactor-http-nio-4] DEBUG reactor.netty.test - [id: 0xe1345d96, L:/0:0:0:0:0:0:0:1:57198 - R:/0:0:0:0:0:0:0:1:57199] READ: 145B POST /test/World HTTP/1.1\r\nContent-Type: text/plain\r\nuser-agent: ReactorNetty/dev\r\nhost: [::1]:57198\r\naccept: */*\r\ntransfer-encoding: chunked\r\n\r\n
```

if we do not replace these chars it will be like this

```
09:39:31.758 [reactor-http-nio-3] DEBUG reactor.netty.test - [id: 0xbb41af44, L:/0:0:0:0:0:0:0:1:57833 - R:/0:0:0:0:0:0:0:1:57828] WRITE: 145B POST /test/World HTTP/1.1
Content-Type: text/plain
user-agent: ReactorNetty/dev
host: [::1]:57828
accept: */*
transfer-encoding: chunked


```

and the idea is this format to be human readable, so the second one is better from readability point of view
wdyt?

Reply:
I think we should stick to `msg.toString(charset)` for now and watch what they will do in the future. Unfortunately `io.netty.buffer.ByteBufUtil#decodeString` is package-private so we can't call it directly.

I removed `.replace("\n", "\\n").replace("\r", "\\r")`. I intentionally added it to "minimize" the output. Both ways are okay for me so I removed it.